src/Phase/Phase.cs

```diff
@@ -16,14 +16,14 @@ public sealed class Phase : ITenantContext
         private readonly SemaphoreSlim _lock = new SemaphoreSlim(1, 1);
         private PhaseState _currentState = new VacantState();
 
-        internal Phase(DependencyResolver resolver, IEventsProvider eventsProvider, Func<string, IDictionary<string, string>> tenantKeysFactory)
+        internal Phase(DependencyResolver resolver, IEventsProvider eventsProvider, Func<string, IDictionary<string, string>> tenantKeysFactory, EventPublisher publisher = null)
         {
             DependencyResolver = resolver;
             EventsProvider = eventsProvider;
             Mediator = new Mediator(resolver);
             Session = new Session(resolver, EventsProvider);
             resolver._session = Session;
-            Publisher = new EventPublisher(resolver);
+            Publisher = publisher ?? new EventPublisher(resolver);
         }
 
         public DependencyResolver DependencyResolver { get; }
```

Comment:
Also, in general, but not specific to this PR, what's with this big goddy `DependencyResolver` guy? It feels like a bit of an antipattern to pass him about and ask him for All The Things. Why not just leave that more directly up to the DI infrastructure? If there are some things where you care about more about the scope, lifecycle or construction of a dependency, just pass in a much more specific `IThingICareAboutFactory` that can help you manage those things.

Reply:
Dependency resolver is basically an abstraction to divorce the framework from any particular IoC container (i.e. ninject, autofac, etc). The approach I have here is taken from what i have seen in other frameworks such as Web API, MVC, and even ServiceStack. I agree we could probably segregate some of the operations into more specific interfaces. This is a good evolution of this object. When I started, I wanted to keep this stupid simple until I figured out how much responsibility this thing actually had. Now that its clear all things it must do we can factor this out in a more SOLID way.

One thing i want to be careful about though is having so many interfaces it becomes a nightmare to figure out how to instantiate everything needed to make phase work. This is one of the big problems i feel what preceeded this framework. So we will need to find a good balance here.